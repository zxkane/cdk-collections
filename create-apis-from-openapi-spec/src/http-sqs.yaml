openapi: "3.0.3"
info:
  title: "httpapi-to-sqs"
  version: "v1.0.0"
  license:
    name: "Apache 2.0"
    url: "https://www.apache.org/licenses/LICENSE-2.0.html"  
servers:
- url: ""
  variables:
    basePath:
      default: ""
paths:
  /i:
    post:
      description: "Writing sessions, metrics and events"
      operationId: "sendEvents"
      parameters:
      - name: "content-type"
        in: "header"
        required: true
        schema:
          type: "string"
      requestBody:
        content:
          application/x-www-form-urlencoded:
            schema:
              $ref: "#/components/schemas/events"
        required: true
      responses:
        "404":
          description: "404 response"
          headers:
            Content-Type:
              schema:
                type: "string"
          content: {}
        "200":
          description: "200 response"
          headers:
            Content-Type:
              schema:
                type: "string"
            Access-Control-Allow-Origin:
              schema:
                type: "string"
          content: {}
        "400":
          description: "400 response"
          headers:
            Content-Type:
              schema:
                type: "string"
          content: {}
        "500":
          description: "500 response"
          headers:
            Content-Type:
              schema:
                type: "string"
          content: {}
      x-amazon-apigateway-request-validator: "requestvalidator"
      x-amazon-apigateway-integration:
        credentials: "{{integrationRoleArn}}"
        integrationSubtype: "SQS-SendMessage"
        requestParameters:
          MessageAttributes: >-
            {
              "sourceIp":{
                "StringValue":"${context.identity.sourceIp}",
                "DataType":"String"
              },
              "apiTimestamp":{
                "StringValue":"${context.requestTimeEpoch}",
                "DataType":"Number"
              }
            }
          MessageBody: "$request.body"
          QueueUrl: "{{queueUrl}}"
        type: "aws_proxy"
        connectionType: "INTERNET"
        payloadFormatVersion: "1.0"
        timeoutInMillis: 2000
components:
  schemas:
    events:
      type: object
      required:
        - app_key
        - device_id
      properties:
        app_key:
          type: string
          example: d7d3f96f594f26acd566577a1501e3c82c7ba8fc
        device_id:
          type: string
          example: b452f74e-4770-48f9-b850-eac57240d547
x-amazon-apigateway-request-validators:
  requestvalidator:
    validateRequestParameters: true
    validateRequestBody: false
x-amazon-apigateway-cors:
  allowOrigins:
    - '*'
  maxAge: 3600
  allowMethods:
    - POST
    - OPTIONS
  exposeHeaders:
    - x-apigateway-header
    - x-amz-date
    - content-type
  allowHeaders:
    - x-apigateway-header
    - x-amz-date
    - content-type