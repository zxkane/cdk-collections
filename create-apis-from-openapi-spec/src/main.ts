import * as fs from 'fs';
import * as path from 'path';
import { App, Stack, StackProps, Duration, RemovalPolicy, CfnOutput } from 'aws-cdk-lib';
import { MethodLoggingLevel, ApiDefinition, SpecRestApi } from 'aws-cdk-lib/aws-apigateway';
import { CfnStage, CfnApi } from 'aws-cdk-lib/aws-apigatewayv2';
import { Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { Queue, QueueEncryption } from 'aws-cdk-lib/aws-sqs';
import { Construct } from 'constructs';
import * as Mustache from 'mustache';
/* eslint @typescript-eslint/no-require-imports: "off" */
const yaml = require('js-yaml');

export class MyStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    // define resources here...
    const bufferQueue = new Queue(this, 'event-queue', {
      encryption: QueueEncryption.SQS_MANAGED,
      visibilityTimeout: Duration.seconds(30),
      removalPolicy: RemovalPolicy.DESTROY,
    });

    const apiRole = new Role(this, 'api-gateway-role', {
      assumedBy: new ServicePrincipal('apigateway.amazonaws.com'),
    });
    bufferQueue.grantSendMessages(apiRole);

    // import openapi as http api
    const variables = {
      integrationRoleArn: apiRole.roleArn,
      queueName: bufferQueue.queueName,
      queueUrl: bufferQueue.queueUrl,
    };
    const openAPISpec = this.resolve(yaml.load(Mustache.render(
      fs.readFileSync(path.join(__dirname, './http-sqs.yaml'), 'utf-8'), variables)));

    const httpApi = new CfnApi(this, 'http-api-to-sqs', {
      body: openAPISpec,
      failOnWarnings: false,
    });

    new CfnStage(this, 'DefaultStage', {
      apiId: httpApi.ref,
      stageName: '$default',
      autoDeploy: true,
    });
    new CfnOutput(this, 'HttpAPIEndpoint', {
      value: httpApi.attrApiEndpoint,
      description: 'url of http api',
    });

    // import openapi as REST api
    const deployOptions = {
      stageName: '',
      loggingLevel: MethodLoggingLevel.ERROR,
      dataTraceEnabled: false,
      metricsEnabled: true,
      tracingEnabled: false,
    };
    const restOpenAPISpec = this.resolve(Mustache.render(
      fs.readFileSync(path.join(__dirname, './rest-sqs.yaml'), 'utf-8'),
      variables));
    new SpecRestApi(this, 'rest-to-sqs', {
      apiDefinition: ApiDefinition.fromInline(restOpenAPISpec),
      endpointExportName: 'APIEndpoint',
      deployOptions,
    });
  }
}

const app = new App();

new MyStack(app, 'create-apis-from-openapi-spec');

app.synth();