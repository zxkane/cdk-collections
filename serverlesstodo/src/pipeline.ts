import { Stack, StackProps, Stage, StageProps } from 'aws-cdk-lib';
import { Cache, LocalCacheMode, BuildSpec } from 'aws-cdk-lib/aws-codebuild';
import { CodePipeline, ShellStep, CodePipelineSource } from 'aws-cdk-lib/pipelines';
import { Construct } from 'constructs';
import { TODOStack } from './main';

class TodolistApplication extends Stage {
  constructor(scope: Construct, id: string, props?: StageProps) {
    super(scope, id, props);

    new TODOStack(this, 'serverlesstodo');
  }
}

export class TodolistPipelineStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // const repo
    const connectArn = scope.node.tryGetContext('SourceConnectionArn');
    if (!connectArn) {throw new Error('Must specify the arn of source repo connection.');}
    const oidcSecret: string = scope.node.tryGetContext('OIDCSerectArn');
    if (!oidcSecret) {throw new Error('Must specify the context "OIDCSerectArn" for storing secret.');}

    const pipeline = new CodePipeline(this, 'Pipeline', {
      synth: new ShellStep('Synth', {
        input: CodePipelineSource.connection('zxkane/cdk-collections', 'master', {
          connectionArn: connectArn,
          codeBuildCloneOutput: true,
        }),
        installCommands: [
          'git submodule init && git submodule update && git submodule sync',
          'npm i --prefix serverlesstodo/frontend',
          'npm run build --prefix serverlesstodo/frontend',
          'yarn --cwd serverlesstodo install --check-files --frozen-lockfile',
        ],
        commands: [
          'cd serverlesstodo',
          'npx projen',
          'npx projen test',
          `npx cdk synth serverlesstodo -c OIDCSerectArn=${oidcSecret} -c SourceConnectionArn=${connectArn} -c CognitoDomainPrefix=todolist-userpool-prod`,
        ],
        primaryOutputDirectory: 'serverlesstodo/cdk.out/',
      }),
      dockerEnabledForSynth: true,
      codeBuildDefaults: {
        cache: Cache.local(LocalCacheMode.SOURCE, LocalCacheMode.DOCKER_LAYER),
      },
      synthCodeBuildDefaults: {
        partialBuildSpec: BuildSpec.fromObject({
          version: '0.2',
          phases: {
            install: {
              'runtime-versions': {
                nodejs: 14,
              },
            },
          },
        }),
      },
    });

    pipeline.addStage(new TodolistApplication(this, 'Prod', {
      env: {
        account: process.env.CDK_DEFAULT_ACCOUNT,
        region: process.env.CDK_DEFAULT_REGION,
      },
    }));
  }
}