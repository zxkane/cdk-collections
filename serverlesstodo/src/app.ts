import { App } from 'aws-cdk-lib';
import { TODOStack } from './main';
import { TodolistPipelineStack } from './pipeline';

const app = new App();

new TodolistPipelineStack(app, 'TodolistPipelineStack', {
  env: {
    account: process.env.CDK_DEFAULT_ACCOUNT,
    region: process.env.CDK_DEFAULT_REGION,
  },
});

new TODOStack(app, 'serverlesstodo', {
  env: {
    account: process.env.CDK_DEFAULT_ACCOUNT,
    region: process.env.CDK_DEFAULT_REGION,
  },
});

app.synth();