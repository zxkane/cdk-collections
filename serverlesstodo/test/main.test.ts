import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { TODOStack } from '../src/main';

test('Snapshot', () => {
  const app = new App();
  const stack = new TODOStack(app, 'test');

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});