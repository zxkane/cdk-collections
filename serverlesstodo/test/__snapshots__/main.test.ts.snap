// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Outputs": Object {
    "TODOAPI": Object {
      "Description": "url of TODO endpoint",
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "todoappbackendapi8D2202DC",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "todoappbackendapiDeploymentStageprod7F1D5431",
            },
            "/todo",
          ],
        ],
      },
    },
    "todoappbackendapiEndpoint8BBE7B48": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "todoappbackendapi8D2202DC",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "todoappbackendapiDeploymentStageprod7F1D5431",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "RequestValidatorB6FDBF18": Object {
      "Properties": Object {
        "RestApiId": Object {
          "Ref": "todoappbackendapi8D2202DC",
        },
        "ValidateRequestBody": true,
        "ValidateRequestParameters": true,
      },
      "Type": "AWS::ApiGateway::RequestValidator",
    },
    "TODOTable8589412C": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "ContributorInsightsSpecification": Object {
          "Enabled": true,
        },
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": true,
        },
        "SSESpecification": Object {
          "SSEEnabled": true,
        },
        "StreamSpecification": Object {
          "StreamViewType": "NEW_AND_OLD_IMAGES",
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "apigatewayroleE6D48DBD": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "dynamodb:UpdateItem",
                    "dynamodb:Scan",
                    "dynamodb:DeleteItem",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "TODOTable8589412C",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "table",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "todoappaccesslogs17D6476B": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "RetentionInDays": 30,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "todoappbackendapi8D2202DC": Object {
      "Properties": Object {
        "Name": "todo-app-backend-api",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "todoappbackendapiAccountCD5BA139": Object {
      "DeletionPolicy": "Retain",
      "DependsOn": Array [
        "todoappbackendapi8D2202DC",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "todoappbackendapiCloudWatchRole367FAC02",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "todoappbackendapiCloudWatchRole367FAC02": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "todoappbackendapiDeployment0F74DA5677ac922f4265c36bcdf9473d76d3528d": Object {
      "DependsOn": Array [
        "RequestValidatorB6FDBF18",
        "todoappbackendapitodotodoIdDELETE0F39924D",
        "todoappbackendapitodotodoIdPOST275E728C",
        "todoappbackendapitodotodoId35AD3DC0",
        "todoappbackendapitodoGET50F454FD",
        "todoappbackendapitodoPUT3C8BE992",
        "todoappbackendapitodo214A50F3",
        "todoappbackendapitodomodel8B506559",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "todoappbackendapi8D2202DC",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "todoappbackendapiDeploymentStageprod7F1D5431": Object {
      "DependsOn": Array [
        "todoappbackendapiAccountCD5BA139",
      ],
      "Properties": Object {
        "AccessLogSetting": Object {
          "DestinationArn": Object {
            "Fn::GetAtt": Array [
              "todoappaccesslogs17D6476B",
              "Arn",
            ],
          },
          "Format": "{\\"requestId\\":\\"$context.requestId\\",\\"ip\\":\\"$context.identity.sourceIp\\",\\"user\\":\\"$context.identity.user\\",\\"caller\\":\\"$context.identity.caller\\",\\"requestTime\\":\\"$context.requestTime\\",\\"httpMethod\\":\\"$context.httpMethod\\",\\"resourcePath\\":\\"$context.resourcePath\\",\\"status\\":\\"$context.status\\",\\"protocol\\":\\"$context.protocol\\",\\"responseLength\\":\\"$context.responseLength\\"}",
        },
        "DeploymentId": Object {
          "Ref": "todoappbackendapiDeployment0F74DA5677ac922f4265c36bcdf9473d76d3528d",
        },
        "MethodSettings": Array [
          Object {
            "DataTraceEnabled": true,
            "HttpMethod": "*",
            "LoggingLevel": "INFO",
            "MetricsEnabled": true,
            "ResourcePath": "/*",
          },
        ],
        "RestApiId": Object {
          "Ref": "todoappbackendapi8D2202DC",
        },
        "StageName": "prod",
        "TracingEnabled": true,
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "todoappbackendapitodo214A50F3": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "todoappbackendapi8D2202DC",
            "RootResourceId",
          ],
        },
        "PathPart": "todo",
        "RestApiId": Object {
          "Ref": "todoappbackendapi8D2202DC",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "todoappbackendapitodoGET50F454FD": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": Object {
          "Credentials": Object {
            "Fn::GetAtt": Array [
              "apigatewayroleE6D48DBD",
              "Arn",
            ],
          },
          "IntegrationHttpMethod": "POST",
          "IntegrationResponses": Array [
            Object {
              "ResponseTemplates": Object {
                "application/json": "#set($inputRoot = $input.path('$'))
[
  #foreach($item in $inputRoot.Items)
  {
    \\"id\\": \\"$item.id.S\\",
    \\"subject\\": \\"$util.escapeJavaScript($item.subject.S)\\",
    \\"description\\": \\"$util.escapeJavaScript($item.description.S)\\",
    \\"dueDate\\": $item.dueDate.N,
    \\"createdTimeInMills\\": $item.createdTimeInMills.N
    #if(\\"$!item.lastModifiedTimeInMills.N\\" != \\"\\")
    ,
    \\"lastModifiedTimeInMills\\": $item.lastModifiedTimeInMills.N
    #end
  }
  #if($foreach.hasNext),#end
  #end
]
",
              },
              "StatusCode": "200",
            },
            Object {
              "ResponseTemplates": Object {
                "text/html": "Error",
              },
              "SelectionPattern": "500",
              "StatusCode": "500",
            },
          ],
          "RequestParameters": Object {
            "integration.request.header.Content-Type": "'application/json'",
          },
          "RequestTemplates": Object {
            "application/json": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "{
  \\"TableName\\": \\"",
                  Object {
                    "Ref": "TODOTable8589412C",
                  },
                  "\\",
  \\"ProjectionExpression\\": \\"id, subject, description, dueDate, createdTimeInMills, lastModifiedTimeInMills\\",
  \\"FilterExpression\\": \\"begins_with(id, :id)\\",
  \\"ExpressionAttributeValues\\": {
      \\":id\\": {\\"S\\": \\"todo-\\"}
  }
}
",
                ],
              ],
            },
          },
          "Type": "AWS",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":dynamodb:action/Scan",
              ],
            ],
          },
        },
        "MethodResponses": Array [
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "200",
          },
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "400",
          },
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "404",
          },
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "500",
          },
        ],
        "OperationName": "Get TODOs",
        "RequestParameters": Object {
          "method.request.header.content-type": true,
        },
        "RequestValidatorId": Object {
          "Ref": "RequestValidatorB6FDBF18",
        },
        "ResourceId": Object {
          "Ref": "todoappbackendapitodo214A50F3",
        },
        "RestApiId": Object {
          "Ref": "todoappbackendapi8D2202DC",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "todoappbackendapitodoPUT3C8BE992": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "PUT",
        "Integration": Object {
          "Credentials": Object {
            "Fn::GetAtt": Array [
              "apigatewayroleE6D48DBD",
              "Arn",
            ],
          },
          "IntegrationHttpMethod": "POST",
          "IntegrationResponses": Array [
            Object {
              "ResponseTemplates": Object {
                "application/json": "{
  \\"subject\\": \\"$util.escapeJavaScript($input.path('$.Attributes.subject.S')).replaceAll(\\"\\\\\\\\'\\",\\"'\\")\\",
  \\"description\\": \\"$util.escapeJavaScript($input.path('$.Attributes.description.S')).replaceAll(\\"\\\\\\\\'\\",\\"'\\")\\",
  \\"id\\": \\"$input.path('$.Attributes.id.S')\\",
  \\"createdTimeInMills\\": $input.path('$.Attributes.createdTimeInMills.N'),
  \\"dueDate\\": $input.path('$.Attributes.dueDate.N')
}
",
              },
              "StatusCode": "200",
            },
            Object {
              "ResponseTemplates": Object {
                "application/json": "#if($input.path('$.__type') == \\"com.amazonaws.dynamodb.v20120810#ConditionalCheckFailedException\\")
{
  \\"message\\": \\"the todo id already exists.\\"
}
#end",
              },
              "SelectionPattern": "4\\\\d{2}",
              "StatusCode": "400",
            },
            Object {
              "ResponseTemplates": Object {
                "text/html": "Error",
              },
              "SelectionPattern": "500",
              "StatusCode": "500",
            },
          ],
          "RequestParameters": Object {
            "integration.request.header.Content-Type": "'application/json'",
          },
          "RequestTemplates": Object {
            "application/json": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "#set($dueDate = $input.path('$.dueDate'))
#set($subject = \\"$util.escapeJavaScript($input.path('$.subject'))\\")
#set($description = \\"$util.escapeJavaScript($input.path('$.description'))\\")
{
  \\"TableName\\": \\"",
                  Object {
                    "Ref": "TODOTable8589412C",
                  },
                  "\\",
  \\"Key\\": {
    \\"id\\": {
        \\"S\\": \\"todo-$context.requestId\\"
      }
  },
  \\"UpdateExpression\\": \\"set subject = :s, description = :d, dueDate = :dd, createdTimeInMills = :ct\\",
  \\"ExpressionAttributeValues\\": {
      \\":s\\": {
        \\"S\\": \\"$subject\\"
      },
      \\":d\\": {
        \\"S\\": \\"$description\\"
      },
      \\":dd\\": {
        \\"N\\": \\"$dueDate\\"
      },
      \\":ct\\": {
        \\"N\\": \\"$context.requestTimeEpoch\\"
      }
  },
  \\"ConditionExpression\\": \\"attribute_not_exists(id)\\",
  \\"ReturnValues\\": \\"ALL_NEW\\"
}
",
                ],
              ],
            },
          },
          "Type": "AWS",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":dynamodb:action/UpdateItem",
              ],
            ],
          },
        },
        "MethodResponses": Array [
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "200",
          },
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "400",
          },
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "404",
          },
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "500",
          },
        ],
        "OperationName": "Create TODO item",
        "RequestModels": Object {
          "application/json": Object {
            "Ref": "todoappbackendapitodomodel8B506559",
          },
        },
        "RequestParameters": Object {
          "method.request.header.content-type": true,
        },
        "RequestValidatorId": Object {
          "Ref": "RequestValidatorB6FDBF18",
        },
        "ResourceId": Object {
          "Ref": "todoappbackendapitodo214A50F3",
        },
        "RestApiId": Object {
          "Ref": "todoappbackendapi8D2202DC",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "todoappbackendapitodomodel8B506559": Object {
      "Properties": Object {
        "ContentType": "application/json",
        "RestApiId": Object {
          "Ref": "todoappbackendapi8D2202DC",
        },
        "Schema": Object {
          "$schema": "http://json-schema.org/draft-04/schema#",
          "properties": Object {
            "description": Object {
              "type": "string",
            },
            "dueDate": Object {
              "type": "number",
            },
            "subject": Object {
              "type": "string",
            },
          },
          "required": Array [
            "subject",
            "description",
            "dueDate",
          ],
          "type": "object",
        },
      },
      "Type": "AWS::ApiGateway::Model",
    },
    "todoappbackendapitodotodoId35AD3DC0": Object {
      "Properties": Object {
        "ParentId": Object {
          "Ref": "todoappbackendapitodo214A50F3",
        },
        "PathPart": "{todoId}",
        "RestApiId": Object {
          "Ref": "todoappbackendapi8D2202DC",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "todoappbackendapitodotodoIdDELETE0F39924D": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "DELETE",
        "Integration": Object {
          "Credentials": Object {
            "Fn::GetAtt": Array [
              "apigatewayroleE6D48DBD",
              "Arn",
            ],
          },
          "IntegrationHttpMethod": "POST",
          "IntegrationResponses": Array [
            Object {
              "SelectionPattern": "200",
              "StatusCode": "200",
            },
            Object {
              "ResponseTemplates": Object {
                "application/json": "#if($input.path('$.__type') == \\"com.amazonaws.dynamodb.v20120810#ConditionalCheckFailedException\\")
{
  \\"message\\": \\"the TODO item does not exist or is not owned by you.\\"
}
#end",
              },
              "SelectionPattern": "400",
              "StatusCode": "400",
            },
            Object {
              "ResponseTemplates": Object {
                "text/html": "Error",
              },
              "SelectionPattern": "500",
              "StatusCode": "500",
            },
          ],
          "RequestParameters": Object {
            "integration.request.header.Content-Type": "'application/json'",
          },
          "RequestTemplates": Object {
            "application/json": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "#set($todoId = \\"$util.escapeJavaScript($input.params('todoId'))\\")  
{
  \\"TableName\\": \\"",
                  Object {
                    "Ref": "TODOTable8589412C",
                  },
                  "\\",
  \\"Key\\": {
    \\"id\\": {
        \\"S\\": \\"todo-$todoId\\"
      }
  }
}
",
                ],
              ],
            },
          },
          "Type": "AWS",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":dynamodb:action/DeleteItem",
              ],
            ],
          },
        },
        "MethodResponses": Array [
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "200",
          },
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "400",
          },
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "404",
          },
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "500",
          },
        ],
        "OperationName": "Delete a TODO item",
        "RequestParameters": Object {
          "method.request.header.content-type": true,
        },
        "RequestValidatorId": Object {
          "Ref": "RequestValidatorB6FDBF18",
        },
        "ResourceId": Object {
          "Ref": "todoappbackendapitodotodoId35AD3DC0",
        },
        "RestApiId": Object {
          "Ref": "todoappbackendapi8D2202DC",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "todoappbackendapitodotodoIdPOST275E728C": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": Object {
          "Credentials": Object {
            "Fn::GetAtt": Array [
              "apigatewayroleE6D48DBD",
              "Arn",
            ],
          },
          "IntegrationHttpMethod": "POST",
          "IntegrationResponses": Array [
            Object {
              "ResponseTemplates": Object {
                "application/json": "{
  \\"subject\\": \\"$util.escapeJavaScript($input.path('$.Attributes.subject.S'))\\",
  \\"id\\": \\"$input.path('$.Attributes.id.S')\\",
  \\"description\\": \\"$util.escapeJavaScript($input.path('$.Attributes.description.S'))\\",
  \\"dueDate\\": $input.path('$.Attributes.dueDate.N')\\",
  \\"createdTimeInMills\\": $input.path('$.Attributes.createdTimeInMills.N'),
  \\"lastModifiedTimeInMills\\": $input.path('$.Attributes.lastModifiedTimeInMills.N'),  
}
",
              },
              "StatusCode": "200",
            },
            Object {
              "ResponseTemplates": Object {
                "application/json": "#if($input.path('$.__type') == \\"com.amazonaws.dynamodb.v20120810#ConditionalCheckFailedException\\")
{
  \\"message\\": \\"the TODO does NOT exists.\\"
}
#end",
              },
              "SelectionPattern": "4\\\\d{2}",
              "StatusCode": "400",
            },
            Object {
              "ResponseTemplates": Object {
                "text/html": "Error",
              },
              "SelectionPattern": "500",
              "StatusCode": "500",
            },
          ],
          "RequestParameters": Object {
            "integration.request.header.Content-Type": "'application/json'",
          },
          "RequestTemplates": Object {
            "application/json": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "#set($dueDate = $input.path('$.dueDate'))
#set($subject = \\"$util.escapeJavaScript($input.path('$.subject'))\\")
#set($description = \\"$util.escapeJavaScript($input.path('$.description'))\\")
#set($todoId = \\"$util.escapeJavaScript($input.params('todoId'))\\")  
{
  \\"TableName\\": \\"",
                  Object {
                    "Ref": "TODOTable8589412C",
                  },
                  "\\",
  \\"Key\\": {
    \\"id\\": {
        \\"S\\": \\"todo-$todoId\\"
      }
  },
  \\"UpdateExpression\\": \\"set subject = :s, description = :d, dueDate = :dd, lastModifiedTimeInMills = :ct\\",
  \\"ExpressionAttributeValues\\": {
      \\":s\\": {
        \\"S\\": \\"$subject\\"
      },
      \\":d\\": {
        \\"S\\": \\"$description\\"
      },
      \\":dd\\": {
        \\"N\\": \\"$dueDate\\"
      },
      \\":ct\\": {
        \\"N\\": \\"$context.requestTimeEpoch\\"
      }
  },
  \\"ConditionExpression\\": \\"attribute_exists(id)\\",
  \\"ReturnValues\\": \\"ALL_NEW\\"
}
",
                ],
              ],
            },
          },
          "Type": "AWS",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":dynamodb:action/UpdateItem",
              ],
            ],
          },
        },
        "MethodResponses": Array [
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "200",
          },
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "400",
          },
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "404",
          },
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "500",
          },
        ],
        "OperationName": "Update TODO item",
        "RequestModels": Object {
          "application/json": Object {
            "Ref": "todoappbackendapitodomodel8B506559",
          },
        },
        "RequestParameters": Object {
          "method.request.header.content-type": true,
        },
        "RequestValidatorId": Object {
          "Ref": "RequestValidatorB6FDBF18",
        },
        "ResourceId": Object {
          "Ref": "todoappbackendapitodotodoId35AD3DC0",
        },
        "RestApiId": Object {
          "Ref": "todoappbackendapi8D2202DC",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
